lib/Greenspun/lib/Dicts.cpp

```diff
@@ -0,0 +1,365 @@
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Lars Maier
+////////////////////////////////////////////////////////////////////////////////
+
+#include <list>
+
+#include <velocypack/Collection.h>
+#include <velocypack/velocypack-aliases.h>
+#include "Dicts.h"
+
+#include "Greenspun/Extractor.h"
+#include "Greenspun/Interpreter.h"
+
+using namespace arangodb::greenspun;
+
+template <bool ignoreMissing>
+EvalResult Prim_DictExtract(Machine& ctx, VPackSlice const paramsList, VPackBuilder& result) {
+  if (paramsList.length() < 1) {
+    return EvalError("expected at least on parameter");
+  }
+  VPackArrayIterator iter(paramsList);
+
+  VPackSlice obj = *iter;
+  if (!obj.isObject()) {
+    return EvalError("expected first parameter to be a dict, found: " + obj.toJson());
+  }
+  ++iter;
+
+  {
+    VPackObjectBuilder ob(&result);
+    for (; iter.valid(); ++iter) {
+      VPackSlice key = *iter;
+      if (!key.isString()) {
+        return EvalError("expected string, found: " + key.toJson());
+      }
+
+      VPackSlice value = obj.get(key.stringRef());
+      if (value.isNone()) {
+        if constexpr (ignoreMissing) {
+          continue;
+        } else {
+          return EvalError("key `" + key.copyString() + "` not found");
+        }
+      }
+
+      result.add(key.stringRef(), value);
+    }
+  }
+  return {};
+}
+
+void createPaths(std::list<std::list<std::string>>& finalPaths,
+                 VPackSlice object, std::list<std::string>& currentPath) {
+  for (VPackObjectIterator iter(object); iter.valid(); iter++) {
+    std::string currentKey = iter.key().toString();
+    VPackSlice currentValue = iter.value();
+
+    if (currentValue.isObject()) {
+      // path not done yet
+      currentPath.emplace_back(std::move(currentKey));
+
+      std::list<std::string> currentTmpPath(currentPath);
+      finalPaths.emplace_back(std::move(currentTmpPath));
+      createPaths(finalPaths, currentValue, currentPath);
+    } else {
+      // path is done
+      std::list<std::string> currentTmpPath(currentPath);
+      currentTmpPath.emplace_back(std::move(currentKey));
+      finalPaths.emplace_back(std::move(currentTmpPath));
+    }
+
+    if (iter.isLast()) {
+      // if end of path reached, remove last visited member
+      if (currentPath.size() > 0) {
+        currentPath.pop_back();
+      }
+    }
+  }
+}
+
+void pathToBuilder(std::list<std::list<std::string>>& finalPaths, VPackBuilder& result) {
+  result.openArray();
+  for (auto const& path : finalPaths) {
+    if (path.size() > 1) {
+      result.openArray();
+    }
+    for (auto const& pathElement : path) {
+      result.add(VPackValue(pathElement));
+    }
+    if (path.size() > 1) {
+      result.close();
+    }
```

Comment:
We're expecting also things like: 
```
let vertexKeysToInsertB = ["a", ["a", "b"], ["a", "c"], "d", "e"];
```

Reply:
What do you mean by "expect"? That'd be the output, not the input. And what I wanted to say is that I'd rather have `[["a"], ["a", "b"], ["a", "c"], ["d"], ["e"]]` as output, because I don't have to make multiple cases when handling that. If someone has a non-recursive dict and wants the keys, there's `dict-keys` for that.

Reply:
I understand. I'll create a ticket for further discussion: https://team-gordo.atlassian.net/browse/TG-1132